File: internal/notifications/events.go
package notifications

import "fmt"

type EventType string

const (
	EventDoseDue        EventType = "DOSE_DUE"
	EventRefillReminder EventType = "REFILL_REMINDER"
	EventTestReminder   EventType = "TEST_REMINDER"
)

type EventPayload struct {
	Event   EventType         `json:"event"`
	To      string            `json:"to"`
	Payload map[string]string `json:"payload"` // e.g. {"patientName": "...", "time": "..."}

	// TTS preferences coming from caregiver UI
	Voice        string  `json:"voice,omitempty"`        // "Charon", "Kore", etc
	Emotion      string  `json:"emotion,omitempty"`      // "calm", "friendly", "urgent"
	SpeakingRate float32 `json:"speakingRate,omitempty"` // 0.8, 1.0, etc
	Prompt       string  `json:"prompt,omitempty"`       // extra style instructions
}

// RenderBody builds a message body from the event type and payload
func RenderBody(ev EventPayload) (string, error) {
	switch ev.Event {
	case EventDoseDue:
		return fmt.Sprintf(
			"Hi %s, this is your DoseDock reminder to take your %s at %s.",
			ev.Payload["patientName"],
			ev.Payload["meds"],
			ev.Payload["time"],
		), nil

	case EventRefillReminder:
		return fmt.Sprintf(
			"Hi %s, your DoseDock dispenser is running low on %s. Please refill soon.",
			ev.Payload["patientName"],
			ev.Payload["meds"],
		), nil

	case EventTestReminder:
		return "DoseDock test reminder, notifications are working.", nil

	default:
		return "", fmt.Errorf("unknown event type %s", ev.Event)
	}
}

File: server/src/notifications/events.ts
import type { EventPayload } from "./types";

function requireField(payload: Record<string, string>, key: string, event: string): string {
  const v = payload?.[key];
  if (!v) throw new Error(`missing payload field '${key}' for event ${event}`);
  return v;
}

export function renderBody(ev: EventPayload): string {
  if (!ev.to) throw new Error("missing 'to'");

  switch (ev.event) {
    case "DOSE_DUE": {
      const patientName = requireField(ev.payload, "patientName", "DOSE_DUE");
      const meds = requireField(ev.payload, "meds", "DOSE_DUE");
      const time = requireField(ev.payload, "time", "DOSE_DUE");
      return `Hi ${patientName}, this is your DoseDock reminder to take your ${meds} at ${time}.`;
    }

    case "REFILL_REMINDER": {
      const patientName = requireField(ev.payload, "patientName", "REFILL_REMINDER");
      const meds = requireField(ev.payload, "meds", "REFILL_REMINDER");
      return `Hi ${patientName}, your DoseDock dispenser is running low on ${meds}. Please refill soon.`;
    }

    case "TEST_REMINDER":
      return "DoseDock test reminder, notifications are working.";

    default: {
      const _exhaustive: never = ev.event;
      return _exhaustive;
    }
  }
}
